Auto-saving when exiting the editor

0nko opened the pull request:
Fixes #4240

To test:
1. Open or create a new post
2. Add changes
3. Navigate from the editor using the back or home button
4. A snackbar pops up, informing the user how the post was saved

There are a few different scenarios which come into play and each shows a different dialog:
1. No changes = no message
2. Changes made to a published post - post saved locally
3. Changes made to a new post - post saved online as a draft
4. Changes made to a draft - post updated online
5. Exiting from an empty post - if post is new, post is discarded, otherwise saved locally
6. Device is offline - post is saved locally
7. Draft post not modified but has local changes - save changes online

hypest commented:
Hey @0nko , please have a look at my comment here: https://github.com/wordpress-mobile/WordPress-Android/issues/4240#issuecomment-229348761

For quick reference: I think we should not present a user with a "save" dialog and we should just save locally on "back" or "home" pressed.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostSettingsFragment.java:
```diff
@@ -343,6 +343,24 @@ public boolean onTouch(View view, MotionEvent motionEvent) {
             mPasswordEditText.setText(mPost.getPassword());
         }
 
+        updateStatusSpinner();
```
Is this change relevant anymore? Not sure but, it might be left over from previous versions of this PR.

Same for changes in `PostUtils.java` and `WPAlertDialogFragment.java`.

0nko replied:
Got rid of it.

0nko replied:
As we've talked, this is necessary for new posts.

hypest commented:
I also tried the "No changes = no message" case with an online draft and while it's true that no message comes up, the post still gets a progress bar on the posts list which leaves me wondering if something got saved or not. Not a deal-breaker but perhaps something we can address as part of this PR, WDYT? Thanks!

0nko commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListActivity.java:
```diff
@@ -53,6 +55,21 @@ public void onResume() {
+                if (data.getBooleanExtra(EditPostActivity.EXTRA_SAVED_AS_LOCAL_DRAFT, false)) {
+                    Snackbar.make(findViewById(R.id.coordinator), R.string.editor_post_saved_locally, Snackbar.LENGTH_LONG).show();
+                } else {
+                    Snackbar.make(findViewById(R.id.coordinator), R.string.editor_post_saved_online, Snackbar.LENGTH_LONG).show();
```
I've updated the code as per your suggestions.

0nko commented:
Stefanos, I believe I addressed all issues. I added a Publish button to the snackbar if a post is publishable. The delay should also be gone.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListFragment.java:
```diff
@@ -46,6 +46,8 @@
 
 import de.greenrobot.event.EventBus;
 
+import static android.app.Activity.RESULT_OK;
```
I think that we could follow the rest of the codebase and just use `Activity.RESULT_OK` where needed and import `Activity` instead of a static import of the constant. WDYT?

0nko replied:
Yup, thanks.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListFragment.java:
```diff
@@ -123,6 +125,51 @@ public void onClick(View v) {
+                } else {
+                    if (isPublishable) {
+                        if (savedLocally) {
+                            Snackbar.make(getActivity().findViewById(R.id.coordinator), R.string.editor_post_saved_locally_not_published, Snackbar.LENGTH_LONG)
```
Let's wrap the long lines (this one and line 151) to 120char as per [the style guide](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/CODESTYLE.md). Thanks!

0nko replied:
Done.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -915,23 +916,67 @@ protected Boolean doInBackground(Void... params) {
+
+            if (mPost.getStatusEnum() != PostStatus.PUBLISHED && mPost.isPublishable()) {
+                publishPost();
+            }
```
Let's merge the `}` and the `else` in a single line as per [the Android style guide](https://source.android.com/source/code-style.html#use-standard-brace-style). Thanks!

0nko replied:
Done.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostUtils.java:
```diff
@@ -127,4 +132,17 @@ public static void trackSavePostAnalytics(Post post) {
         }
     }
+
+    public static void showCustomDialog(Activity activity, String title, String message,
```
This helper doesn't seem to be used anywhere. Is it redundant maybe?

0nko commented on WordPress/src/main/res/values/strings.xml:
```diff
@@ -957,6 +959,10 @@
     <string name="editor_page_title_placeholder">Page Title</string>
     <string name="editor_content_placeholder">Share your story here…</string>
+    <string name="editor_post_saved_locally">The post was saved on your device</string>
+    <string name="editor_post_saved_locally_not_published">The post was saved on your device, but wasn\'t published.</string>
```
Sounds better :).

0nko commented on WordPress/src/main/res/values/strings.xml:
```diff
@@ -1077,8 +1083,8 @@
-    <string name="error_publish_empty_post">Can\'t publish an empty post</string>
-    <string name="error_publish_no_network">Can\'t publish while there is no connection. Saved as draft.</string>
+    <string name="error_publish_empty_post">Can\'t publish an empty post.</string>
+    <string name="error_publish_no_network">Can\'t publish while there is no connection. Changes saved on the device.</string>
```
Good idea! I've updated it like you suggested.

0nko commented:
I've updated the code as per your suggestions and it's ready for yet another pass. 

I had to fix a couple of build issues because the FluxC changes are now merged and I've used the old code. It all seems to work fine. Note that I've added one more scenario (7) to the PR description.

hypest commented on WordPress/src/main/res/values/strings.xml:
```diff
@@ -940,9 +940,9 @@
     <string name="editor_content_placeholder">Share your story here…</string>
     <string name="editor_post_saved_locally">The post was saved on your device</string>
-    <string name="editor_post_saved_locally_not_published">The post was saved on your device, but wasn\'t published.</string>
+    <string name="editor_post_saved_locally_not_published">The post saved locally only. Consider publishing it.</string>
```
Hmm, this one seems weird grammatically. Perhaps we can lose the `The` from the string start. Thanks!

aforcier commented:
Noticed a crash when attempting to publish an empty post:

```
java.lang.NullPointerException: Attempt to invoke virtual method 'boolean android.os.Handler.post(java.lang.Runnable)' on a null object reference
	  at org.wordpress.android.ui.posts.EditPostActivity$5.run(EditPostActivity.java:858)
	  at java.lang.Thread.run(Thread.java:761)
```

aforcier commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostSettingsFragment.java:
```diff
@@ -74,6 +74,7 @@
 import org.wordpress.android.util.DisplayUtils;
 import org.wordpress.android.util.EditTextUtils;
 import org.wordpress.android.util.GeocoderUtils;
+import org.wordpress.android.util.JSONUtils;
```
This import is unused (probably came in with the merge conflict).

0nko replied:
Removed.

aforcier commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostUtils.java:
```diff
@@ -1,5 +1,9 @@
+import android.app.Activity;
+import android.app.FragmentManager;
+
+import org.wordpress.android.R;
```
This import is also unused.

0nko replied:
Removed.

aforcier commented on WordPress/src/main/res/values/strings.xml:
```diff
@@ -1059,8 +1065,8 @@
     <string name="error_edit_comment">An error occurred while editing the comment</string>
-    <string name="error_publish_empty_post">Can\'t publish an empty post</string>
-    <string name="error_publish_no_network">Can\'t publish while there is no connection. Saved as draft.</string>
+    <string name="error_publish_empty_post">Can\'t publish an empty post.</string>
```
To be consistent with the rest of our messaging (and Android style guides), I think we should leave this without a period.

0nko replied:
Fixed.

aforcier commented:
@0nko this looks good 👍 - I'll give it one final run once the crash issue is fixed.

0nko commented:
While I was fixing the crash I've discovered that even though the back button auto-saving works, the Publish and Save buttons don't behave as expected. There also was a lot of duplicate code, so I've refactored it and it should now be working properly.

@hypest,@aforcier: Would you mind giving it another look? It'd be great if you could try different saving and publishing scenarios, just to make sure I haven't missed something.

Thanks!

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -1221,27 +1207,36 @@ private void savePostAndFinish() {
-        if (isNewPost() && !isPublishable) {
-            mDispatcher.dispatch(PostActionBuilder.newRemovePostAction(mPost));
-        }
+    private boolean updatePostObject() {
```
Hey @0nko ! This version of `updatePostObject` is quite different to the `updatePostObject(boolean)` one.

* It calculates and returns a boolean but doesn't use it internally
* The other version of updatePostObject doesn't return a value at all

Because of these, I would suggest splitting this into a `boolean isFirstTimePublish()` and a `void updatePostObject()` for clearer functionality.

Also, it seems that it catches the `IllegalEditorStateException`, thus allowing its caller to continue normally, without a user notification. Not sure at this point but, can you confirm that this is indeed the intended behaviour here? The other version of `updatePostObject` does throw it. Thanks!

0nko replied:
> Because of these, I would suggest splitting this into a boolean isFirstTimePublish() and a void updatePostObject() for clearer functionality.

Good point.

> Also, it seems that it catches the IllegalEditorStateException, thus allowing its caller to continue normally, without a user notification. Not sure at this point but, can you confirm that this is indeed the intended behaviour here? The other version of updatePostObject does throw it. Thanks!

You're right. It doesn't make sense to continue if the post update fails. I fixed the flow to account for this.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -1187,23 +1154,111 @@ protected Boolean doInBackground(Void... params) {
+
+        boolean postUpdateSuccessful = updatePostObject();
+        if (!postUpdateSuccessful) {
+            return;
```
It seems that currently, `updatePostObject()` can fail only when the `EditorFragment.java` fragment is not added to the activity. I think it's OK to silently return in this case.

Since this is not exactly obvious by looking at the code here, may I suggest adding a small comment about it? Something along the lines of "just return since the only case updatePostObject() can fail is when the editor fragment not being added to the activity". WDYT?

0nko replied:
I'll add it, thanks

aforcier commented:
Gave this another pass, looks good, all the cases listed in the OP are working and didn't run into any issues. 👍
